File: tide_graph.rb
# tide_graph.rb
AMPLITUDE = 1.5
MEAN_LEVEL = 0.5
PERIOD = 12.42

def tide_level(hours)
  rad = (hours / PERIOD) * 2 * Math::PI
  MEAN_LEVEL + AMPLITUDE * Math.sin(rad - 1.2)
end

def generate_data
  data = []
  (0...24*60).step(10) do |minutes|
    h = minutes / 60.0
    lvl = tide_level(h)
    data << [h, lvl]
  end
  data
end

def find_extrema(data)
  highs = []
  lows = []
  (1...data.length-1).each do |i|
    prev, curr, nxt = data[i-1], data[i], data[i+1]
    if curr[1] > prev[1] && curr[1] > nxt[1]
      highs << curr
    elsif curr[1] < prev[1] && curr[1] < nxt[1]
      lows << curr
    end
  end
  [highs, lows]
end

def draw_graph(data)
  levels = data.map { |d| d[1] }
  min_lvl = levels.min - 0.2
  max_lvl = levels.max + 0.2

  height = 20
  scale = (max_lvl - min_lvl) / height

  cols = 50
  step = data.length / cols
  graph_data = []
  cols.times do |i|
    idx = i * step
    h, lvl = data[idx]
    row = ((max_lvl - lvl) / scale).round
    row = 0 if row < 0
    row = height - 1 if row >= height
    graph_data << { h: h, lvl: lvl, row: row }
  end

  lines = []
  lines << "🌊 Tide Graph (24‑hour forecast)"
  lines << ""
  lines << "Level (m)"

  height.times do |r|
    lvl = max_lvl - r * scale
    line = format("%5.1f |", lvl)
    cols.times do |c|
      line << (graph_data[c][:row] == r ? "*" : " ")
    end
    lines << line
  end

  lines << "      +" + "-" * cols

  time_line = "       "
  (0..24).step(3) do |i|
    pos = (i / 24.0 * cols).to_i
    label = format("%02d:00", i)
    time_line << " " * (pos - time_line.length) if time_line.length < pos
    time_line << label
  end
  lines << time_line
  lines << ""

  highs, lows = find_extrema(data)
  unless highs.empty?
    str = "High tides: "
    highs.first(2).each_with_index do |h, idx|
      str << ", " if idx > 0
      hour = h[0].to_i
      min = ((h[0] - hour) * 60).round
      str << format("%02d:%02d (%.1fm)", hour, min, h[1])
    end
    lines << str
  end
  unless lows.empty?
    str = "Low tides:  "
    lows.first(2).each_with_index do |l, idx|
      str << ", " if idx > 0
      hour = l[0].to_i
      min = ((l[0] - hour) * 60).round
      str << format("%02d:%02d (%.1fm)", hour, min, l[1])
    end
    lines << str
  end

  lines.join("\n")
end

puts draw_graph(generate_data)

File: tide_graph.py
# tide_graph.py
import math
import datetime

# Tide model parameters (semi‑diurnal, M2 + S2 approximation)
AMPLITUDE = 1.5
MEAN_LEVEL = 0.5
PERIOD_HOURS = 12.42  # M2 period

def tide_level(hours):
    """Return tide level in meters for a given hour of the day."""
    # Two tidal constituents: M2 (principal lunar) and S2 (solar)
    # Phase offset to make high tides around 02:00 and 14:30
    phase = 0.0  # we'll adjust with time
    rad = (hours / PERIOD_HOURS) * 2 * math.pi
    level = MEAN_LEVEL + AMPLITUDE * math.sin(rad - 1.2)
    return level

def generate_data():
    """Generate (hour, level) pairs for 24 hours, sampled every 10 minutes."""
    data = []
    for minutes in range(0, 24*60, 10):
        h = minutes / 60.0
        level = tide_level(h)
        data.append((h, level))
    return data

def find_extrema(data):
    """Find high and low tide times and levels."""
    highs, lows = [], []
    for i in range(1, len(data)-1):
        prev, curr, nxt = data[i-1], data[i], data[i+1]
        if curr[1] > prev[1] and curr[1] > nxt[1]:
            highs.append(curr)
        elif curr[1] < prev[1] and curr[1] < nxt[1]:
            lows.append(curr)
    return highs, lows

def draw_graph(data):
    """Draw an ASCII graph of the tide data."""
    # Determine min and max level
    levels = [d[1] for d in data]
    min_lvl, max_lvl = min(levels), max(levels)
    # Add some padding
    min_lvl = min(round(min_lvl - 0.2, 1), -1.5)
    max_lvl = max(round(max_lvl + 0.2, 1), 1.5)

    # Scale to fit graph height (20 rows)
    height = 20
    scale = (max_lvl - min_lvl) / height

    # Create graph grid (rows = levels, columns = time points)
    # We'll use 50 columns for time (one per ~30 minutes)
    cols = 50
    step = len(data) // cols
    graph_data = []
    for i in range(cols):
        idx = i * step
        h, lvl = data[idx]
        # Map level to row (0 = top, height-1 = bottom)
        row = int(round((max_lvl - lvl) / scale))
        if row < 0: row = 0
        if row >= height: row = height - 1
        graph_data.append((h, lvl, row))

    # Build the graph string
    lines = []
    # Title
    lines.append("🌊 Tide Graph (24‑hour forecast)")
    lines.append("")
    # Y-axis label
    lines.append("Level (m)")

    # For each row
    for r in range(height):
        lvl = max_lvl - r * scale
        line = f"{lvl:5.1f} |"
        # Fill columns
        for col in range(cols):
            if graph_data[col][2] == r:
                # Use a star for the tide line
                line += "*"
            else:
                # Use space, but add a dot for zero level?
                line += " "
        lines.append(line)

    # X-axis
    x_axis = "      +" + "-" * cols
    lines.append(x_axis)
    # Time labels
    time_labels = ""
    for i in range(0, 25, 3):
        label = f"{i:02d}:00"
        # Position label under the graph (approx)
        pos = int(i / 24 * cols)
        # We'll just print a separate line with labels
    # Actually easier: print a time line
    time_line = "       "
    for i in range(0, 25, 3):
        pos = int(i / 24 * cols)
        time_line += " " * (pos - len(time_line)) + f"{i:02d}:00"
    lines.append(time_line)
    lines.append("")

    # Find extrema
    highs, lows = find_extrema(data)
    if highs:
        high_str = ", ".join([f"{int(h[0]):02d}:{int((h[0]%1)*60):02d} ({h[1]:.1f}m)" for h in highs[:2]])
        lines.append(f"High tides: {high_str}")
    if lows:
        low_str = ", ".join([f"{int(l[0]):02d}:{int((l[0]%1)*60):02d} ({l[1]:.1f}m)" for l in lows[:2]])
        lines.append(f"Low tides:  {low_str}")

    return "\n".join(lines)

def main():
    data = generate_data()
    graph = draw_graph(data)
    print(graph)

if __name__ == "__main__":
    main()
